src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java

```diff
@@ -0,0 +1,318 @@
+                }
+                final String taxId = tokens[0];
+                final String parent = tokens[1];
+                final String rank = tokens[2];
+                final PSPathogenReferenceTaxonProperties taxonProperties;
+                if (taxIdToProperties.containsKey(taxId)) {
+                    taxonProperties = taxIdToProperties.get(taxId);
+                } else {
+                    taxonProperties = new PSPathogenReferenceTaxonProperties();
+                    taxonProperties.name = "tax_" + taxId;
+                    taxIdsNotFound.add(taxId);
+                }
+                taxonProperties.rank = rank;
+                if (!taxId.equals(ROOT_ID)) { //keep root's parent set to null
+                    taxonProperties.parentTaxId = parent;
+                }
+                taxIdToProperties.put(taxId, taxonProperties);
+            }
+            return taxIdsNotFound;
+        } catch (final IOException e) {
+            throw new UserException.CouldNotReadInputFile("Error reading from taxonomy dump nodes file", e);
+        }
+    }
+
+    /**
+     * Helper function for building the map from tax id to reference contig accession
+     */
+    private static void updateAccessionTaxonProperties(final String accession, final String name, final long length, final Map<String, PSPathogenReferenceTaxonProperties> taxIdToProperties) {
+        final PSPathogenReferenceTaxonProperties taxonProperties;
+        if (taxIdToProperties.containsKey(accession)) {
+            taxonProperties = taxIdToProperties.get(accession);
+        } else {
+            taxonProperties = new PSPathogenReferenceTaxonProperties();
+        }
+        taxonProperties.accessions.add(name);
+        taxonProperties.length += length;
+        taxIdToProperties.put(accession, taxonProperties);
+    }
+
+    /**
+     * Create reference_name-to-taxid map (just an inversion on taxIdToProperties)
+     */
+    protected static Map<String, String> buildAccessionToTaxIdMap(final Map<String, PSPathogenReferenceTaxonProperties> taxIdToProperties) {
+        final Map<String, String> accessionToTaxId = new HashMap<>();
+        for (final String taxId : taxIdToProperties.keySet()) {
+            final PSPathogenReferenceTaxonProperties taxonProperties = taxIdToProperties.get(taxId);
+            for (String name : taxonProperties.accessions) {
+                accessionToTaxId.put(name, taxId);
+            }
+        }
+        return accessionToTaxId;
+    }
+
+    /**
+     * Returns a PSTree representing a reduced taxonomic tree containing only taxa present in the reference
+     */
+    protected static PSTree buildTaxonomicTree(final Map<String, PSPathogenReferenceTaxonProperties> taxIdToProperties) {
+
+        //Build tree of all taxa
+        final PSTree tree = new PSTree(ROOT_ID);
+        final Collection<String> invalidIds = new ArrayList<>();
```

Comment:
Changed to a hash set. I'm not sure exactly how many it is typically. I will just initialize it to the total number of ID's since this is not huge (~10,000's).